cmd/lakectl/cmd/fs_cp.go

```diff
@@ -0,0 +1,145 @@
+package cmd
+
+import (
+	"context"
+	"fmt"
+	"github.com/spf13/cobra"
+	"github.com/treeverse/lakefs/pkg/api/apigen"
+	"github.com/treeverse/lakefs/pkg/api/apiutil"
+	"github.com/treeverse/lakefs/pkg/uri"
+	"net/http"
+	"os"
+	"path"
+	"strings"
+	"sync"
+)
+
+var fsCpCmd = &cobra.Command{
+	Use:   "cp <source URI> <destination URI>",
+	Short: "Copy object",
+	Args:  cobra.ExactArgs(2),
+	Run: func(cmd *cobra.Command, args []string) {
+		recursive := Must(cmd.Flags().GetBool(recursiveFlagName))
+		concurrency := Must(cmd.Flags().GetInt("concurrency"))
+		sourceURI := MustParsePathURI("source", args[0])
+		destinationURI := MustParsePathURI("destination", args[1])
+
+		if sourceURI.Repository != destinationURI.Repository {
+			Die("Can only copy files in the same repo", 1)
+		}
+
+		client := getClient()
+
+		ctx := cmd.Context()
+
+		if !recursive {
+			err := copyObject(ctx, client, sourceURI, destinationURI)
+			if err != nil {
+				DieErr(err)
+			}
+			return
+		}
+
+		success := true
+		var errorsWg sync.WaitGroup
+		errors := make(chan error)
+		errorsWg.Add(1)
+		go func() {
+			defer errorsWg.Done()
+			for err := range errors {
+				_, _ = fmt.Fprintln(os.Stderr, err)
+				success = false
+			}
+		}()
+
+		var copyWg sync.WaitGroup
+		paths := make(chan string)
+		copyWg.Add(concurrency)
+
+		for i := 0; i < concurrency; i++ {
+			go copyObjectWorker(ctx, client, sourceURI, destinationURI, paths, errors, &copyWg)
+		}
+
+		prefix := *sourceURI.Path
+		var paramsDelimiter apigen.PaginationDelimiter = ""
+		var from string
+		pfx := apigen.PaginationPrefix(prefix)
+		for {
+			params := &apigen.ListObjectsParams{
+				Prefix:    &pfx,
+				After:     apiutil.Ptr(apigen.PaginationAfter(from)),
+				Delimiter: &paramsDelimiter,
+			}
+			resp, err := client.ListObjectsWithResponse(cmd.Context(), sourceURI.Repository, sourceURI.Ref, params)
+			DieOnErrorOrUnexpectedStatusCode(resp, err, http.StatusOK)
+			if resp.JSON200 == nil {
+				Die("Bad response from server", 1)
+			}
+
+			results := resp.JSON200.Results
+			for i := range results {
+				paths <- results[i].Path
+			}
+
+			pagination := resp.JSON200.Pagination
+			if !pagination.HasMore {
+				break
+			}
+			from = pagination.NextOffset
+		}
+
+		close(paths)
+		copyWg.Wait()
+		close(errors)
```

Comment:
This races with the error-handling goroutine above: once all copiers are done, if the error-handling goroutine is waiting then this line may close errors _before_ that goroutine reads again.  That results in reading on a closed channel, which is a race: the errors goroutine will end up discovering an error, but the code below on l. 95 will see `success==true` and exit cleanly when it should fail.

Refactoring to _load_ from a goroutine and _unload_ from the main thread of execution helps avoid this race.